backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/storage/backup/HybridBackupCommand.java

```diff
@@ -0,0 +1,291 @@
+        if (!FeatureSupported.isHybridBackupSupported(getVm())) {
+            return false;
+        }
+
+        return super.validate();
+    }
+
+    @Override
+    protected void executeCommand() {
+        VmBackup vmBackup = getParameters().getVmBackup();
+
+        Guid toCheckpointId = Guid.newGuid();
+        ActionReturnValue returnValue = runInternalAction(
+                ActionType.CreateSnapshotForVm,
+                buildCreateSnapshotParameters(toCheckpointId),
+                createStepsContext(StepEnum.CREATING_SNAPSHOTS));
+        if (!returnValue.getSucceeded()) {
+            auditLog(AuditLogType.USER_FAILED_CREATE_SNAPSHOT);
+            getReturnValue().setFault(returnValue.getFault());
+            updateVmBackupPhase(VmBackupPhase.FAILED);
+            return;
+        }
+
+        auditLog(AuditLogType.USER_CREATE_SNAPSHOT);
+        Guid snapshotId = returnValue.getActionReturnValue();
+        Guid vmBackupId = createVmBackup(snapshotId);
+        log.info("Created VmBackup entity '{}' for VM '{}'", vmBackupId, vmBackup.getVmId());
+        getParameters().setAutoGeneratedSnapshotId(snapshotId);
+        vmBackup.setDisks(getDisks());
+        getParameters().setToCheckpointId(toCheckpointId);
+        updateVmBackupPhase(VmBackupPhase.STARTING);
+        setActionReturnValue(vmBackupId);
+
+        setSucceeded(true);
+    }
+
+    @Override
+    public boolean performNextOperation(int completedChildCount) {
+        restoreCommandState();
+        if (getParameters().getVmBackup().getPhase() == VmBackupPhase.STARTING) {
+            auditLog(AuditLogType.USER_CREATE_SNAPSHOT_FINISHED_SUCCESS);
+            VmBackup vmBackup = getParameters().getVmBackup();
+            Map<Guid, Guid> diskIdToSnapshot = diskImageDao.getAllSnapshotsForVmSnapshot(getParameters().getAutoGeneratedSnapshotId())
+                    .stream()
+                    .collect(Collectors.toMap(DiskImage::getId, diskImage -> diskImage.getImageId()));
+            TransactionSupport.executeInNewTransaction(() -> {
+                getDisks().forEach(
+                        disk -> {
+                            setDiskBackupModeIfSupported(disk);
+                            vmBackupDao.addDiskToVmBackup(vmBackup.getId(), disk.getId(), diskIdToSnapshot.get(disk.getId()));
+                        });
+
+                return null;
+            });
+
+            Guid toCheckpointId = createVmCheckpoint(getParameters().getToCheckpointId());
+            vmBackup.setToCheckpointId(toCheckpointId);
+
+            updateVmBackupPhase(VmBackupPhase.READY);
+            return true;
+        }
+
+        if (getParameters().getVmBackup().getPhase() == VmBackupPhase.READY) {
+            return true;
+        }
+
+        if (getParameters().getVmBackup().getPhase() == VmBackupPhase.FINALIZING) {
+            removeAutoGeneratedSnapshot(createStepsContext(StepEnum.MERGE_SNAPSHOTS), getActionType(), getParameters());
```

Comment:
@ahadas I think a call to freeLock() should be made here, otherwise endAction might release the lock after RemoveSnapshot releases it

Reply:
so are you concerned about the following flow:
1. HybridBackup acquires the lock
2. CreateSnapshotForVm releases the lock
3. RemoveSnapshot acquires the lock
4. RemoveSnapshot releases the lock
5. HybridBackup attempts to release the lock
and want to call freeLock here in order to drop the lock from the context to prevent step 5?
it's not mandatory - calling release lock when the lock has not been acquired is ok
I'm not concerned about the flow above but about the call to removeAutoGeneratedSnapshot in the endWithFailure method - if remove snapshot would lock the vm and right after HybridBackup then try to release it
so yeah, calling freeLock here or right before calling removeAutoGeneratedSnapshot in endWithFailure makes sense to me

Reply:
The problem is that endAction reconstructs the lock[1], so even if it's null because it was released, it will rebuild it without acquiring, so looks like it should be in both endAction methods

Reply:
yeah, better to unlock it in the endAction then
btw, the lock is supposed to be rebuilt also when the engine restarts

Reply:
right, and it's even worse because it is re-acquired and will block snapshot removal... I'm trying something unorthodox